### DISCUSSION TITLE
Current and Voltage in Sub Circuit Component

### DISCUSSION THREAD

**Discussion by Abbas111977** [SpiceSharp/SpiceSharp#198] [Q&A]
Dears,
Can i measure voltage and current of component inside a sub circuit ? if yes then How ?
Kind regards

**Comment by svenboulanger** [comment]
Hi,

You can get to the behaviors using [this property](https://github.com/SpiceSharp/SpiceSharp/blob/dace7bddaadf7ac22bc32abd2dc61b4d28afb13d/SpiceSharp/Components/Subcircuits/Behaviors/EntitiesBehavior.cs#L40).
There should be an `EntitiesBehavior` in the simulation for each subcircuit.

```csharp
var behaviors = sim.EntityBehaviors["subcktName"].GetProperty<IBehaviorContainerCollection>("behaviors");
double value = behaviors["componentInSubcktName"].GetProperty<double>("myProperty");
```

If you want to measure voltages inside a subcircuit, you will have to expand the name like you often do in Spice simulators. The internal node `A` in subcircuit `X1` would be resolved to `X1/A` by default.
If you are using local solvers for parallellization reasons, you will have to go via the local state doing

```csharp
var p = sim.EntityBehaviors["subcktName"].GetParameterSet<SpiceSharp.Components.Subcircuits.EntitiesBehavior>();
var state = p.GetState<IBiasingSimulationState>();
// Now you can extract your voltages and so on from the local state.
```
